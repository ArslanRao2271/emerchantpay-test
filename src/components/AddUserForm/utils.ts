export interface TestInputField {
  inputField: string;
  textToFind: string;
}

export const inputFields: Array<TestInputField> = [
  {
    inputField: "Email address",
    textToFind: "Email is required",
  },
  {
    inputField: "First Name",
    textToFind: "First name is required",
  },
  {
    inputField: "Last Name",
    textToFind: "Last name is required",
  },
];

export const emailFieldValidation: Array<string> = ["tester", "tester@foxtons"];

export const nameFieldValidation: Array<string> = ["Ar@", "test*"];
