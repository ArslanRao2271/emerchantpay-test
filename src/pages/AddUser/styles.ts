import { SxProps } from "@mui/material";
import { boxDisplay } from "../../appStyles";

const styles: Record<string, SxProps> = {
  addUserPageContainer: {
    ...boxDisplay("flex", "center", "center"),
    flexDirection: "column",
    gap: "5rem",
  },
};

export default styles;
