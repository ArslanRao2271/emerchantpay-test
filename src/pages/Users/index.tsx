// libs
import React from "react";

// src
import { TableComponent } from "@components/TableComponent";

// styles

const headerData = [
  {
    label: "Property Documents",
    width: "40%",
  },
  {
    label: "Uploaded on",
    width: "20%",
  },
  {
    label: "Status",
    width: "20%",
  },
  {
    label: "Action",
  },
];

function Users() {
  return (
    <>
      <TableComponent headerData={headerData} rowsData={[]} />
    </>
  );
}

export default Users;
