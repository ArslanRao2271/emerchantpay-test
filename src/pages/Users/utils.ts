export const users = [
  {
    firstName: "Test1",
    lastName: "test",
    email: "test1@mailinator.com",
  },
  {
    firstName: "Test2",
    lastName: "test",
    email: "test2@mailinator.com",
  },
  {
    firstName: "Test3",
    lastName: "test",
    email: "test3@mailinator.com",
  },
  {
    firstName: "Test4",
    lastName: "test",
    email: "test4@mailinator.com",
  },
  {
    firstName: "Test5",
    lastName: "test",
    email: "test5@mailinator.com",
  },
  {
    firstName: "Test6",
    lastName: "test",
    email: "test6@mailinator.com",
  },
  {
    firstName: "Test7",
    lastName: "test",
    email: "test7@mailinator.com",
  },
  {
    firstName: "Test8",
    lastName: "test",
    email: "test8@mailinator.com",
  },
  {
    firstName: "Test9",
    lastName: "test",
    email: "test9@mailinator.com",
  },
  {
    firstName: "Test10",
    lastName: "test",
    email: "test10@mailinator.com",
  },
];
