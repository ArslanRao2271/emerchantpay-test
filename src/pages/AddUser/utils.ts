import { UserValues } from "@components/AddUserForm";

export const parseUserData = (
  data: UserValues[],
  onDeleteClick?: (email: string) => void,
  handleModalOpen?: ({ email, firstName, lastName }: UserValues) => void
) =>
  data.map(({ firstName, lastName, email }) => ({
    id: email,
    rowData: [
      {
        label: email,
        width: "40%",
        onClick: () => handleModalOpen({ email, firstName, lastName }),
      },
      { label: firstName, width: "20%" },
      { label: lastName, width: "20%" },
      {
        label: "",
        type: "action-buttons",
        actionButtons: [
          {
            buttonLabel: "Delete",
            size: "small",
            variant: "secondary",
            onClick: () => onDeleteClick(email),
          },
        ],
      },
    ],
  }));
