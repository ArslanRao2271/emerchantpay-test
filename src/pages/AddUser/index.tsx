// libs
import { Box, Typography } from "@mui/material";
import React, { useState } from "react";

// src
import { parseUserData } from "./utils";
import { AddUserForm, UserValues } from "@components/AddUserForm";
import { TableComponent } from "@components/TableComponent";
import { Alert, LibAlertProps } from "@components/Alert";

// styles
import styles from "./styles";

const alertInitialValues: LibAlertProps = {
  isOpen: false,
  severity: "success",
  message: "",
};

const headerData = [
  {
    label: "Email",
    width: "40%",
  },
  {
    label: "First Name",
    width: "20%",
  },
  {
    label: "Last Name",
    width: "20%",
  },
  {
    label: "Action",
    width: "20%",
  },
];

function AddUser() {
  const [rowsData, setRowsData] = useState<UserValues[]>([]);
  const [alert, setAlert] = useState(alertInitialValues);

  const handleAddUser = (user: UserValues) => {
    if (rowsData.some(({ email }) => email === user.email)) {
      setAlert({
        isOpen: true,
        message: `${user.email} already exists.`,
        severity: "error",
      });

      return;
    }

    setRowsData((prevState) => [...prevState, user]);
    setAlert({
      isOpen: true,
      message: `${user.email} added successfully.`,
      severity: "success",
    });
  };

  const handleDelete = (val: string) => {
    setRowsData((prevState) => prevState.filter(({ email }) => email !== val));
  };

  return (
    <Box sx={styles.addUserPageContainer} width="100%">
      <Typography variant="h3">Add User</Typography>
      <AddUserForm getUserData={handleAddUser} />
      <TableComponent
        headerData={headerData}
        rowsData={parseUserData(rowsData, (email) => handleDelete(email))}
      />
      <Alert
        {...alert}
        onClose={() =>
          setAlert((prevState) => ({ ...prevState, isOpen: false }))
        }
      />
    </Box>
  );
}

export default AddUser;
