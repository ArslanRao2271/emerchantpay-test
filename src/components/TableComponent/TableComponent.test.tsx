import { render, screen } from "@testing-library/react";
import { TableComponent } from ".";
import { MemoryRouter } from "react-router-dom";
import { headerData } from "../../pages/AddUser";
import { users } from "../../pages/Users/utils";
import { parseUserData } from "../../pages/AddUser/utils";

const getWrapperComponent = (comp: JSX.Element) => {
  return render(comp, { wrapper: MemoryRouter });
};

describe("TableComponent test cases", () => {
  test("Should match previous snapshot", () => {
    const { asFragment } = getWrapperComponent(
      <TableComponent headerData={[]} rowsData={[]} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  headerData.forEach(({ label }) =>
    test(`Should render label ${label}`, () => {
      getWrapperComponent(
        <TableComponent headerData={headerData} rowsData={[]} />
      );

      expect(screen.queryByText(label)).toBeVisible();
    })
  );

  users.forEach(({ email }) =>
    test(`Should render email ${email}`, () => {
      getWrapperComponent(
        <TableComponent
          headerData={headerData}
          rowsData={parseUserData(users)}
        />
      );

      expect(screen.queryByText(email)).toBeVisible();
    })
  );

  users.forEach(({ firstName }) =>
    test(`Should render firstName ${firstName}`, () => {
      getWrapperComponent(
        <TableComponent
          headerData={headerData}
          rowsData={parseUserData(users)}
        />
      );

      expect(screen.queryByText(firstName)).toBeVisible();
    })
  );
});
