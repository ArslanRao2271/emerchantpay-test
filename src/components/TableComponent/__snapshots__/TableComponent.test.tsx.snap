// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableComponent test cases Should match previous snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-7822g1"
  >
    <div
      class="MuiTableContainer-root css-rorn0c-MuiTableContainer-root"
    >
      <table
        class="MuiTable-root css-rqglhn-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          />
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        />
      </table>
    </div>
  </div>
</DocumentFragment>
`;
